### DISCUSSION TITLE
How to extract client extension types?

### DISCUSSION THREAD

**Discussion by isaachinman** [prisma/prisma#19725] [Q&A] [answered]
I am defining a client extension as such:

```tsx
const prismaClient = _prismaClient.$extends({
  query: {
    company: {
      update: async ({ args, query }) => {
        if (args.data?.status === CompanyStatus.DECLINED) {
          args.data.user = {
            update: {
              accountLocked:
                AccountLockedReason.COMPANY_DECLINED,
            },
          }
        }

        return query(args)
      },
    },
  },
})
```

The TypeScript types for `$extends` seem to be working perfectly, although if I have a look at the generated types in `node_modules/.prisma/client/index.d.ts`, the output is extremely complex.

I would like to organise my client extensions in a separate dir, for the sake of maintainability:

```tsx
// prismaClient.ts
const prismaClient = _prismaClient.$extends({
  query: {
    company: companyExtensions,
  },
})

// companyExtensions.ts
export const companyExtensions: NeedsType = {
  update: async ({ args, query }) => {
    if (args.data?.status === CompanyStatus.DECLINED) {
      args.data.user = {
        update: {
          accountLocked:
            AccountLockedReason.COMPANY_DECLINED,
        },
      }
    }

    return query(args)
  },
}

```

I have struggled for over an hour to try and isolate types for a separated `companyExtensions` object. This is as far as I've gotten:

```tsx
type CompanyExtensions = Parameters<typeof prismaClient['$extends']>[0]
```

I know that Prisma has a special `defineExtension` function, but it seems to be targeted towards distribution of generalised client extensions and does not contain specific `arg` typing for my models themselves.

Would appreciate any help that can be given.

**Comment by isaachinman** [comment, marked as answer]
Ended up solving it like this:

```tsx
type ExtensionArgs = Extract<
  Parameters<typeof _prismaClient.$extends>[0],
  { name?: string }
>
```

**Reply by kguirguis** [reply]
Were you able to get it working with model extensions? It seem that when I try to "extend" the model to include a new function, it complains:

```
Object literal may only specify known properties, and 'mySuperCoolFunction' does not exist in type '{} & { [K: symbol]: { ctx: {} & { findFirst: <A extends MyModelFindFirstArgs<InternalArgs & DefaultArgs>>(args?: Exact<A, MyModelFindFirstArgs<InternalArgs & DefaultArgs>> | undefined) => DynamicModelExtensionFluentApi<...> & PrismaPromise<...>; ... 13 more ...; groupBy: <A extends MyModelGroupByArgs<...>>(args: Exact<...>) ...'
```

**Comment by SevInf** [comment]
Hi @isaachinman.
`defineExtension` is absolutely the way splitting extension in a separate file while retaining autocomplete was designed be solved:

```ts
import { Prisma } from '@prisma/client'
export const myExtension = Prisma.defineExtension({
  update: async ({ args, query }) => {
    if (args.data?.status === CompanyStatus.DECLINED) {
      args.data.user = {
        update: {
          accountLocked:
            AccountLockedReason.COMPANY_DECLINED,
        },
      }
    }

    return query(args)
})
```

>and does not contain specific arg typing for my models themselves.

I was not able to reproduce it. If that's the case for you, please open an issue and share the reproduction. Thanks!

**Reply by isaachinman** [reply]
@SevInf Not sure what I was doing before – it does indeed look like `defineExtension` has typings for my models.

However, the type output of `defineExtension` is very poor/generic. Why is it not possible to retain the actual shape of the extension configuration?

Also, do you have an examples of how to write Jest coverage for a `defineExtension` declaration?

It's more verbose, but I have better typings and test coverage with this setup:

```tsx
const enforceTypeAsConst =
  <Constraint
    extends Record<PropertyKey, unknown> | undefined = undefined
  >() => <T extends Constraint>(obj: T): T => obj

export type ExtensionArgs = Extract<
  Parameters<typeof Prisma.defineExtension>[0],
  { name?: string }
>

export const extension = enforceTypeAsConst<ExtensionArgs['query']>()({
  company: {
    update: async ({ args, query }) => {
      if (args.data?.status === CompanyStatus.DECLINED) {
        args.data.user = {
          update: {
            accountLocked:
              AccountLockedReason.COMPANY_DECLINED,
          },
        }
      }

      return query(args)
    },
  },
})
```

**Reply by SevInf** [reply]
Return type is the unfortunate compromise we had to take to make TS  work correctly within both `defineExtension` and the the exteneded client. 

Regarding testing, I'd probably apply extension to a client instance and test it that way rather then picking up individual properties of extension. That would also mean that don't need very specific type for that extension to test it.

**Reply by isaachinman** [reply]
> I'd probably apply extension to a client instance and test it that way

That is a very heavy-handed way to cover what should be dead simple unit tests.

**Reply by isaachinman** [reply]
I'll stick with exact typings and simple objects + unit tests, but I appreciate your responses. Thanks @SevInf!

**Comment by Dakuan** [comment]
![image](https://github.com/prisma/prisma/assets/633621/75b44a40-347e-4e4d-8361-52d772130360)
 my types also not quite doing the trick for defineExtension
